File: lib/core/animal_counts_merge.dart
import '../domain/models/animal_count.dart';
import 'animal_counts_normalizer.dart';

/// Merges [incoming] counts into [existing] per `categoryId`. Entries in
/// [incoming] overwrite the count for matching ids; other categories remain
/// untouched. Result is normalized (sorted, zeros dropped, dedup-summed).
List<AnimalCount> mergeAnimalCountsByCategory(
  List<AnimalCount> existing,
  List<AnimalCount> incoming,
) {
  final byId = <int, int>{
    for (final a in existing) a.categoryId: a.count,
  };
  for (final a in incoming) {
    byId[a.categoryId] = a.count;
  }
  return normalizeAnimalCounts([
    for (final entry in byId.entries)
      AnimalCount(categoryId: entry.key, count: entry.value),
  ]);
}

File: src/lib/animal-counts-merge.ts
import type { AnimalCount } from '@/domain/models/animal-count';

export function mergeAnimalCounts(lists: AnimalCount[][]): AnimalCount[] {
  const order: string[] = [];
  const sums = new Map<string, number>();
  for (const list of lists) {
    for (const { categoryId, count } of list) {
      if (!sums.has(categoryId)) order.push(categoryId);
      sums.set(categoryId, (sums.get(categoryId) ?? 0) + count);
    }
  }
  return order.map((id) => ({ categoryId: id, count: sums.get(id)! }));
}
